### PR TITLE
fix: undeterministic tests

### PR DESCRIPTION
**Pull request by sammy1991106**
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues
- [ ] Tag type of change

### CODE REVIEW

File: src/state.rs [hunk 1]
```diff
@@ -93,29 +93,25 @@ mod tests {
...
-    fn teardown() {
-        let _ = fs::remove_file(PathBuf::from(PATH));
-    }
-
-    fn run_test<T>(test: T)
+    fn run_test<T>(state_path: PathBuf, test: T)
     where
         T: FnOnce() + panic::UnwindSafe,
     {
         let result = panic::catch_unwind(test);
```
Comment (cgorenflo) [inline src/state.rs:100]:
```suggestion
        T: FnOnce(&PathBuf) + panic::UnwindSafe,
    {
        let result = panic::catch_unwind(||test(state_path));
```

File: src/state.rs [hunk 2]
```diff
@@ -93,29 +93,25 @@ mod tests {
...
     use super::State;
 
-    const PATH: &str = "./test_state.json";
-
-    fn teardown() {
-        let _ = fs::remove_file(PathBuf::from(PATH));
-    }
-
-    fn run_test<T>(test: T)
+    fn run_test<T>(state_path: PathBuf, test: T)
```
Comment (cgorenflo) [inline src/state.rs:96]:
```suggestion
    fn run_test<T>(state_path: &PathBuf, test: T)
```

File: src/state.rs [hunk 3]
```diff
@@ -93,29 +93,25 @@ mod tests {
...
     }
 
     #[test]
     fn new_state_should_read_from_the_file() {
-        run_test(|| {
-            let state = State::new(PATH.into()).unwrap();
+        let state_path = PathBuf::from("./new_state_should_read_from_the_file.json");
+
+        run_test(state_path.clone(), move || {
+            let state = State::new(state_path.clone()).unwrap();
```
Comment (cgorenflo) [inline src/state.rs:110]: state actually only needs a reference to PathBuf, I don't think it needs to own it

File: src/state.rs (hunk 2 above)
Comment (sammy1991106) [reply src/state.rs:96]: ya of course, I didn't even think much of it lol.

File: src/lib.rs [hunk 4]
```diff
@@ -26,25 +26,25 @@ mod url;
...
     event_processor: EventProcessor,
     event_processor_driver: event_processor::EventProcessorDriver,
     state_updater: state::Updater,
-    state: State,
+    state: State<'a>,
 }
 
-impl App {
-    pub fn new(tm_client: tendermint_rpc::HttpClient, state: State) -> Self {
+impl<'a> App<'a> {
```
Comment (cgorenflo) [inline src/lib.rs:46]:
```suggestion
impl App<'_> {
```
I think this should work, then you don't have to specify the lifetime in the fns anymore

File: src/state.rs [hunk 5]
```diff
@@ -21,13 +21,13 @@ pub enum Error {
...
-pub struct State {
+pub struct State<'a> {
     state: HashMap<String, block::Height>,
-    path: PathBuf,
+    path: &'a PathBuf,
 }
 
-impl State {
-    pub fn new(path: PathBuf) -> Result<Self, Error> {
+impl<'a> State<'a> {
```
Comment (cgorenflo) [inline src/state.rs:29]:
```suggestion
impl State<'_> {
```

File: src/state.rs [hunk 6]
```diff
@@ -93,29 +93,25 @@ mod tests {
...
-            let state = State::new(PATH.into()).unwrap();
+        let state_path = PathBuf::from("./new_state_should_read_from_the_file.json");
+
+        run_test(&state_path, || {
+            let state = State::new(&state_path).unwrap();
             assert_eq!(state.state.len(), 0);
 
-            fs::write(PathBuf::from(PATH), String::from("{\"a\": \"2\", \"b\": \"3\"}")).unwrap();
-            let state = State::new(PATH.into()).unwrap();
+            fs::write(state_path.clone(), String::from("{\"a\": \"2\", \"b\": \"3\"}")).unwrap();
```
Comment (cgorenflo) [inline src/state.rs:113]:
```suggestion
            fs::write(&state_path, String::from("{\"a\": \"2\", \"b\": \"3\"}")).unwrap();
```

**Review by cgorenflo — APPROVED** [review]

**Pull request merged**